contracts/meta-staking/src/msg.rs

```diff
@@ -1,11 +1,82 @@
 use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{
+    AllDelegationsResponse, AllValidatorsResponse, BondedDenomResponse, Coin, Decimal,
+    DelegationResponse, ValidatorResponse,
+};
+
+use crate::state::Config;
 
 #[cw_serde]
-pub struct InstantiateMsg {}
+pub struct InstantiateMsg {
+    pub local_denom: String,
+    pub provider_denom: String,
+    pub consumer_provider_exchange_rate: Decimal,
+}
 
 #[cw_serde]
-pub enum ExecuteMsg {}
+pub enum ExecuteMsg {
+    /// This is translated to a [MsgDelegate](https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/cosmos/staking/v1beta1/tx.proto#L81-L90).
+    /// `delegator_address` is automatically filled with the current contract's address.
+    Delegate { validator: String, amount: Coin },
+    /// This is translated to a [MsgUndelegate](https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/cosmos/staking/v1beta1/tx.proto#L112-L121).
+    /// `delegator_address` is automatically filled with the current contract's address.
+    Undelegate { validator: String, amount: Coin },
+    /// This is translated to a [[MsgWithdrawDelegatorReward](https://github.com/cosmos/cosmos-sdk/blob/v0.42.4/proto/cosmos/distribution/v1beta1/tx.proto#L42-L50).
+    /// `delegator_address` is automatically filled with the current contract's address.
+    WithdrawDelegatorReward {
+        /// The `validator_address`
+        validator: String,
+    },
+}
 
 #[cw_serde]
 #[derive(QueryResponses)]
-pub enum QueryMsg {}
+pub enum QueryMsg {
+    /// Returns the denomination that can be bonded (if there are multiple native tokens on the chain)
+    #[returns(BondedDenomResponse)]
+    BondedDenom {},
+    /// AllDelegations will return all delegations by the delegator
+    #[returns(AllDelegationsResponse)]
+    AllDelegations { delegator: String },
+    /// Returns meta-staking config
+    #[returns(Config)]
+    Config {},
+    /// Delegation will return more detailed info on a particular
+    /// delegation, defined by delegator/validator pair
+    #[returns(DelegationResponse)]
+    Delegation {
+        delegator: String,
+        validator: String,
+    },
+    /// Returns all validators in the currently active validator set.
+    ///
+    /// The query response type is `AllValidatorsResponse`.
+    #[returns(AllValidatorsResponse)]
+    AllValidators {},
+    /// Returns the validator at the given address. Returns None if the validator is
+    /// not part of the currently active validator set.
+    ///
+    /// The query response type is `ValidatorResponse`.
+    #[returns(ValidatorResponse)]
+    Validator {
+        /// The validator's address (e.g. (e.g. cosmosvaloper1...))
+        address: String,
+    },
+}
+
+#[cw_serde]
+pub enum SudoMsg {
```

Comment:
Agree with this design... but, let's use that with gov contract (set in init), which can be dao or simply an account we control for demo.

I would make
```rust
enum ExecuteMsg {
  Sudo(SudoMsg),
}
```

to wrap this and perform check before dispatching. Otherwise, we can keep the implementation almost the same

Reply:
(actually we can do it both ways)